File: Realize/6Week/PracticeGraph1.cpp
#include <iostream>
#include <vector>

using namespace std;

class MyGraphMatrix {
private:
    vector<char> vertices;
    vector<vector<int>> adjMat;
    int elemCnt;
    int size;

public:
    MyGraphMatrix(int size) : size(size), elemCnt(0) {
        vertices.resize(size);
        adjMat.resize(size, vector<int>(size, 0));
    }

    bool isFull() const {
        return elemCnt == size;
    }

    void addVertex(char data) {
        if (isFull()) {
            cout << "Graph is Full!!" << endl;
            return;
        }
        vertices[elemCnt++] = data;
    }

    void addEdge(int x, int y) {
        adjMat[x][y] = 1;
        adjMat[y][x] = 1;
    }

    void addDirectedEdge(int x, int y) {
        adjMat[x][y] = 1;
    }

    void deleteEdge(int x, int y) {
        adjMat[x][y] = 0;
        adjMat[y][x] = 0;
    }

    void deleteDirectedEdge(int x, int y) {
        adjMat[x][y] = 0;
    }

    void printAdjacentMatrix() const {
        cout << "  ";
        for (int i = 0; i < elemCnt; ++i) {
            cout << vertices[i] << " ";
        }
        cout << endl;

        for (int i = 0; i < elemCnt; ++i) {
            cout << vertices[i] << " ";
            for (int j = 0; j < elemCnt; ++j) {
                cout << adjMat[i][j] << " ";
            }
            cout << endl;
        }
    }
};

int main() {
    MyGraphMatrix graph(4);
    graph.addVertex('A');
    graph.addVertex('B');
    graph.addVertex('C');
    graph.addVertex('D');

    graph.addEdge(0, 1);
    graph.addEdge(0, 2);
    graph.addEdge(1, 2);
    graph.addEdge(1, 3);
    graph.addEdge(2, 3);
    graph.printAdjacentMatrix();

    return 0;
}

File: Realize/6Week/PracticeGraph1.py
class MyGraphMatrix:
    def __init__(self, size):
        self.vertices = [None] * size
        self.adj_mat = [[0] * size for _ in range(size)]
        self.elem_cnt = 0
        self.size = size

    def is_full(self):
        return self.elem_cnt == self.size

    def add_vertex(self, data):
        if self.is_full():
            print("Graph is Full!!")
            return
        self.vertices[self.elem_cnt] = data
        self.elem_cnt += 1

    def add_edge(self, x, y):
        self.adj_mat[x][y] = 1
        self.adj_mat[y][x] = 1

    def add_directed_edge(self, x, y):
        self.adj_mat[x][y] = 1

    def delete_edge(self, x, y):
        self.adj_mat[x][y] = 0
        self.adj_mat[y][x] = 0

    def delete_directed_edge(self, x, y):
        self.adj_mat[x][y] = 0

    def print_adjacent_matrix(self):
        print("  ", end="")
        for i in range(self.elem_cnt):
            print(self.vertices[i], end=" ")
        print()

        for i in range(self.elem_cnt):
            print(self.vertices[i], end=" ")
            for j in range(self.elem_cnt):
                print(self.adj_mat[i][j], end=" ")
            print()

# Test
graph = MyGraphMatrix(4)
graph.add_vertex('A')
graph.add_vertex('B')
graph.add_vertex('C')
graph.add_vertex('D')

graph.add_edge(0, 1)
graph.add_edge(0, 2)
graph.add_edge(1, 2)
graph.add_edge(1, 3)
graph.add_edge(2, 3)
graph.print_adjacent_matrix()
